### PR TITLE
Show diff in error msg when repr is long.

### PR DESCRIPTION
**Pull request by CharString**
Cleaned up #18 into a single patch.

### CODE REVIEW

**Comment by CharString** [comment]
Added a bugfix; expecting bytes in py3 would raise exceptions.

**Comment by garybernhardt** [comment]
Seems like a good change. My only worry is the py3k bytes test. It's big with lots of complex data in it, and the short circuiting conditional is vexing me. :)

With this change, expecter only tries to diff either known string types or the output of pformat. I don't think that the bytes-specific test is necessary as long as we can assume:
1. pformat will return a string
2. difflib won't blow up when given strings

Those seem like safe assumptions to me. If so, I think that the bytes test can just be deleted. But if I'm missing something (very possible), fill me in!

File: expecter.py [hunk 1]
```diff
@@ -7,7 +8,7 @@
     import __builtin__ as __builtins__
 
 
-basestring = getattr(__builtins__, 'basestring', (str, bytes))
+basestring = getattr(__builtins__, 'basestring', str)
```
Comment (sigmavirus24) [inline expecter.py:11]: I'm confused, why are you removing bytes?

Comment (CharString) [inline expecter.py:11]: In python 2 str usage is mixed... Sometimes it's contains a naive string in the same encoding as the source file, sometimes it contains bytes. And then there's the unicode type.

In python 3 str is really just a unicode string type that contain codepoints and bytes really are just bytes. So to treat them as strings in difflib we must either guess the encoding or face `TypeError: string argument without an encoding` as  a2671f07c6055420c02e5c1e61db02e761f6d858 shows.

Now pformat doesn't guess the encoding bytes, but shows the bytevalues that could be interpreted in ASCII as chars, and the rest as two nibbles in hex escaped by `\x` (aka calling repr on a byte)... Calling repr on a byte makes much more sense.

Perhaps a even more sense would be showing just 12 pairs of nibbles separated by spaces on a line, like hexeditors do. And passing _that_ to difflib.

**Comment by CharString** [comment]
@garybernhardt, the shortcircuit  on the message reeks and could be removed... Or we could change message for bytes in py3 into something like

```
Expected other bytes
Diff:
@@ -1 +1 @@
- 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 00
+ 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f
```

But then again. When you're dealing with bytes, chances are you are using something like structlib in your expectations anyway, to keep you sane.

**Comment by CharString** [comment]
Unless it's the 1st of April _and_ actual contains b'[\xf0\x9f\x92\xa9](http://codepoints.net/U+1F4A9)', then I feel inclined that the bytes are in fact utf-8 encoded.

**Comment by sigmavirus24** [comment]
> When you're dealing with bytes, chances are you are using something like structlib in your expectations anyway, to keep you sane.

That's a pretty heavy assumption and one which frankly doesn't float with me.

**Comment by garybernhardt** [comment]
For this PR, I think it's fine to just diff the bytes' repr and remove the bytes-specific test. If bytes repeatedly become an issue in practice, that can be a separate PR. I'd rather avoid implementing such a thing preemptively. :)

**Comment by CharString** [comment]
@garybernhardt should I reduce the test to

```
def it_can_compare_bytes(self):
    null = bytes((0,))
    expect(null) == null
    data = bytes(range(9, 32))
    def _fails():
        expect(data) == data + null
    assert_raises(AssertionError, _fails)
```

to prevent a regression? Or should it be removed all together?

**Comment by garybernhardt** [comment]
I'm tempted to say that even that test isn't needed, since `it_shows_diff_for_large_reprs` already tests repr diffing, and we can trust repr and the stdlib. However, the basestring/str/bytes stuff is so error-prone, and there are so many subtle potential failure paths, that leaving a test in is probably a good idea. Push that simplified test and I'll merge this guy.

Unrelated to that, thanks for making that `split('\n')` -> `splitlines()` change. I'm embarrassed at having missed that!

**Comment by CharString** [comment]
Done. It'd be nice if the version on pypi got updated. I noticed the current one there doesn't even pass arguments to custom expectations.

On the `split('\n')`, I was to blame for that one in the first place. :$ 2e3c9058dc31d1a41380d25a8200ef0768dd183b

**Comment by garybernhardt** [comment]
Merged. Thanks!

**Comment by CharString** [comment]
Thanks for sharing this code.

**Comment by garybernhardt** [comment]
Expecter 0.3.0 is pushed to PyPI!

**Comment by CharString** [comment]
On 25 June 2014 20:36:47 CEST, Gary Bernhardt notifications@github.com wrote:

> Expecter 0.3.0 is pushed to PyPI!

W00T!!

Follow Gary's lead, everybody: get your partyhats on!!

(for the e-archaeologists of y3k: it's a reference to his profile picture... now that I got your attention, is python 2.7 still in use???)

**Pull request merged**